### DISCUSSION TITLE
add_text

### DISCUSSION THREAD

**Discussion by nqnga** [highfestiva/finplot#292] [Q&A] [answered]
I had this issue of add_text:

**fplt.add_text((row['projected_time'], 1.2*mid), 'k=%.2f' % row['trend'], color='#993')
TypeError: must be real number, not str**

My time is timestamp but why I had this problem? 
row['projected_time']: timestamp

How should I fix this issue? 

**my add_line only had issue of :
    assert t <= t0, 'must plot this primitive in prior time-range'**

my input data is a dictionary which contain time and trend like this
![image](https://user-images.githubusercontent.com/9681519/149664643-6d02bbc9-2f71-46fc-9761-ceea478ca61f.png)
I want to add text for trend and and vertical lines for time, but it had error 
Thanks!
 this is my code:

    fplt.candlestick_ochl(data[['Open', 'Close', 'High', 'Low']])

    for row in arcana:
        if row['trend']=='up':
            fplt.add_line((pd.to_datetime(row['projected_time']),min_p), (pd.to_datetime(row['projected_time']),mmax_p), color='#50c976',width=2)
        else:
            fplt.add_line((row['projected_time'],min_p), (row['projected_time'],mmax_p), color='#a73062',width=2)
        fplt.add_text((pd.to_datetime(row['projected_time']), 1.2*mid), 'k=%.2f' % row['trend'], color='#993')
        fplt.add_text((pd.to_datetime(row['projected_time']), 1.3*mid), 'k=%.2f' % row['ruler'], color='#993')
        fplt.add_text((pd.to_datetime(row['projected_time']), 1.4*mid), 'k=%.2f' % row['hit_zone'], color='#993')
        fplt.add_text((pd.to_datetime(row['projected_time']), 1.5*mid), 'k=%.2f' % row['inversed to'], color='#993')

    for row in arr_dead_end:
        fplt.add_line((pd.to_datetime(row['projected_time']), 1.5* min_p), (pd.to_datetime(row['projected_time']), mmax_p), color='#0e2f44', width=3)
        fplt.add_text((pd.to_datetime(row['projected_time']), 1.2*mid), 'k=%.2f' % row['ruler'], color='#993')
        fplt.add_text((pd.to_datetime(row['projected_time']), 1.3*mid), 'k=%.2f' % row['trend'], color='#993')

    fplt.show()

**Comment by highfestiva** [comment, marked as answer]
`TypeError: must be real number, not str` is where your problem lies, look no further. It has nothing to do with finplot. Best thing for a n00b to do is to split each operation into a separate line. That way you learn the basics, including how to handle type errors.

**Reply by nqnga** [reply]
Hi, Can I ask how can modify the time zone? I set the timezone is easten/us. However, the chart is still plotted the europe time zone.
Above the data, My input data is only text for add text and time like picture here 
https://user-images.githubusercontent.com/9681519/149664643-6d02bbc9-2f71-46fc-9761-ceea478ca61f.png
It is not complicated data so that can show errors, I just do not know more :D

**Reply by highfestiva** [reply]
See [wiki ](https://github.com/highfestiva/finplot/wiki/Snippets#time-zone).
